src/actions/actionLock.ts

```diff
@@ -0,0 +1,57 @@
+import { newElementWith } from "../element/mutateElement";
+import { CODES, KEYS } from "../keys";
+import { getSelectedElements } from "../scene";
+import { register } from "./register";
+
+export const actionLock = register({
+  name: "lock",
+  perform: (elements, appState) => {
+    const selectedElements = getSelectedElements(elements, appState, false);
+    const selectedElementIds = new Set(
+      selectedElements.map((element) => element.id),
+    );
+
+    if (selectedElementIds.size === 0) {
+      return false;
+    }
+
+    // It's safe to infer the operation from the first element because selection logic will ensure that selected elements are all of the same locked state
+    const operation = selectedElements[0].locked ? "unlock" : "lock";
+
+    return {
+      elements: elements.map((element) => {
+        if (!selectedElementIds.has(element.id)) {
+          return element;
+        }
+
+        return newElementWith(element, { locked: operation === "lock" });
+      }),
+      appState,
+      commitToHistory: true,
+    };
+  },
+  contextItemLabel: (elements, appState) => {
+    const selected = getSelectedElements(elements, appState, false);
+    if (selected.length === 1) {
+      return selected[0].locked ? "labels.lock.unlock" : "labels.lock.lock";
+    }
+
+    if (selected.length > 1) {
+      return selected[0].locked
+        ? "labels.lock.unlockAll"
+        : "labels.lock.lockAll";
+    }
+
+    throw new Error(
+      "Unexpected zero elements to lock. This should never happen.",
+    );
+  },
+  keyTest: (event, appState, elements) => {
+    const selected = getSelectedElements(elements, appState, false);
+    if (selected.length === 0) {
+      return false;
+    }
+
+    return event.code === CODES.K && event[KEYS.CTRL_OR_CMD] && event.shiftKey;
```

Comment:
We can't have `Cmd+L` for the shortcut because that's a browser shortcut (at least in chrome) for focusing the address bar.

`Cmd+Shift+L` is then the next obvious choice and is the one used in tldraw however I couldn't get this to work. I never receive the event for the `L` key press.

Any idea why?

Reply:
How were you matching the key? If using our `KEYS.L`, then that won't work because holding the `Shift` capitalizes it (`event.key` is a printable character). So you'd have to test for capital L (or use `event.code`, but we should use `event.key` whenever possible).

> We can't have `Cmd+L` for the shortcut because that's a browser shortcut (at least in chrome) for focusing the address bar.

Btw, this event is preventable so we could override the browser behavior. Miro does that as they use `Cmd/Ctrl-L` for locking.

Figma/tldraw uses `Cmd-Shift-L`. I suppose we should use it too.

Reply:
I've tried a bunch of different things. The issue is I never receive the event when I press all 3 keys together.

- Hold shift - keyTest is called
- Hold Cmd - keyTest is called
- Press L - keyTest is not called

Reply:
Turned out the 1Password extension was taking that shortcut and blocking the event 😱